volatility3/framework/contexts/__init__.py

```diff
@@ -462,3 +462,149 @@ def __init__(
         Module.__init__(
             self, context, name, layer_name, offset, symbol_table_name, layer_name
         )
+
+
+class MacOSKernelCacheSupportModule(Module):
+    """
+    If the MH_FILESET KernelCache support is ON, header addresses in all mach-o segments and sections
+    of the MH_FILESET are slid by a specific offset. This is problematic, as some kernel symbols, typically
+    contained in an external ISF file, won't be correctly readable with a sole KASLR shift.
+    The original "kernel" module is the "KernelCache" module, but for compatibility reasons
+    it is still named "kernel".
+    To circumvent this, we create an additional module object, in this context, with the
+    "vm_kernel_slide" shift as offset. Doing so, we are able to detect which slide to use,
+    depending on a provided symbol address.
+
+    Additional reference, on the stage where the KernelCache is given an additional slide :
+     - https://github.com/apple-open-source/macos/blob/14.3/xnu/osfmk/i386/i386_init.c#L621
+    """
+
+    def __init__(
+        self, context: interfaces.context.ContextInterface, config_path: str, name: str
+    ) -> None:
+        super().__init__(context, config_path, name)
+
+        pathjoin = interfaces.configuration.path_join
+        context.config[pathjoin(config_path, "vm_kernel_slide")] = context.config[
+            pathjoin(config_path, self.layer_name, "vm_kernel_slide")
+        ]
+        context.config[pathjoin(config_path, "kernel_start")] = (
+            context.config[pathjoin(config_path, self.layer_name, "kernel_start")]
+            & self.context.layers[self.layer_name].address_mask
+        )
+        context.config[pathjoin(config_path, "kernel_end")] = (
```

Comment:
In the case of a KernelCache, the `offset` will be the start of the machO KernelCache __TEXT, and not the "classic" kernel __TEXT start.

Using these parameters allow to use pre-determined values, instead of calculating them "on the fly", risking to mess things up. 

FYI, here is a typical configuration, right before exiting `mac.py` automagic : 

```json
{
  "kernel_banner": "Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:23 PDT 2021; root:xnu-8019.41.5~1/RELEASE_X86_64\u0000",
  "kernel_end": 18446743524345970688,
  "kernel_start": 18446743524335484928,
  "kernel_virtual_offset": 384188416,
  "memory_layer": "VmwareLayer",
  "page_map_offset": 451420160,
  "vm_kernel_slide": 379650048
}
```

Reply:
This is a module, the module takes an offset and a size (or is supposed to).  If it needs additional parameters, that's fine, but from the looks of it, it's just used different parameter names for values that actually mean the same as offset and size.  The offset is where the module appears in virtual memory, and size is how many bytes it is.